js/ssai_heartbeat.js

```diff
@@ -0,0 +1,155 @@
+OO.plugin('heartbeat', function(OO, _, $) {
+  var log = function() {
+    OO.log.apply(this, $.merge(['heartbeat:'], arguments));
+  };
+
+  var heartbeat = function(mb) {
+    var _this = this;
+    var config = {};
+
+    var DEFAULT_CONFIG = Object.freeze({
+      Interval: 10 * 1000,
+      ReportingPathPattern: '<hostname>/v1/vod_playback_pos/<embed_code>?ssai_guid=<ssai_guid>'
+    });
+
+    //constants
+    _this.CONSTANTS = Object.freeze({
+      HA_ASSET_FLAG: 'ha_enabled',
+      HA_IGNORE_MAX_FLAG: 'ha_ignore_max_timeout',
+      DEFAULT_SEGMENT_LENGTH: 4,
+      PROXY_REFRESH_TIME: 6,
+    });
+
+    var streamURL = '';
+    var ssai_guid = '';
+    var embed_code = '';
+
+    var movieDuration = 0;
+    var playheadPosition = 0;
+    var hostname = '//ssai.ooyala.com';
+    var reportingPaused = false;
+
+    var heartbeatTimer = 0;
+
+    initialize();
+
+    function initialize() {
+      config = buildConfig(config);
+
+      //events subscribe
+      mb.subscribe(OO.EVENTS.VC_WILL_PLAY, 'heartbeat', _onVcWillPlay);
+      mb.subscribe(OO.EVENTS.EMBED_CODE_CHANGED, 'heartbeat', _onEmbedCodeChanged);
+      mb.subscribe(OO.EVENTS.PLAYHEAD_TIME_CHANGED, 'heartbeat', _onPlayheadTimeChange);
+      mb.subscribe(OO.EVENTS.PAUSE, 'heartbeat', _onPause);
+      mb.subscribe(OO.EVENTS.PLAY, 'heartbeat', _onPlay);
+      mb.subscribe(OO.EVENTS.VC_PLAYED, 'hearbeat', _onPlayed);
+      mb.subscribe(OO.EVENTS.DESTROY, 'heartbeat', destroy)
+      
+
+      log('initialization completed', config);
+    }
+
+    function stopHeartBeat() {
+      clearInterval(heartbeatTimer);
+    }
+
+    function startHeartBeat() {
+      stopHeartBeat();
+
+      heartbeatTimer = setInterval(reportHeartBeat, config.Interval);
+    }
+
+    function reportHeartBeat() {
+        if(reportingPaused){
```

Comment:
Indentation seems is a bit too deep here, should be only two spaces instead of four

Reply:
done